roxie/topo/toposerver.cpp

```diff
@@ -0,0 +1,325 @@
+                    else
+                        updateTopology(line);
+                }
+                regenerateResponse();
+                bool match = suppliedDigest.append("\n").compare(cachedDigest) == 0;
+                unsigned rlen = cachedDigest.length();
+                if (!match)
+                    rlen += cachedResponse.length();
+                _WINREV(rlen);
+                client->write(&rlen, 4);
+                if (traceLevel>=5)
+                    DBGLOG("Sending digest %s", cachedDigest.str());
+                client->write(cachedDigest.str(), cachedDigest.length());
+                if (!match)
+                {
+                    if (traceLevel>=5)
+                        DBGLOG("Sending response %s", cachedResponse.str());
+                    client->write(cachedResponse.str(), cachedResponse.length());
+                }
+                else if (traceLevel>=5)
+                    DBGLOG("Sending empty response as no changes");
+            }
+            try
+            {
+                client->write("\x0\x0\x0\x0",4);
+            }
+            catch(IException * e)
+            {
+                // Completely ignore an exception here - someone that did not want a reply may have already closed the socket
+                e->Release();
+            }
+            if (traceLevel)
+                reportTopology();
+        }
+        catch(IException * e)
+        {
+            EXCLOG(e, "Failed to process request");
+            e->Release();
+        }
+    }
+    pinger.join();
+}
+
+int main(int argc, const char *argv[])
+{
+    EnableSEHtoExceptionMapping();
+    setTerminateOnSEH();
+    init_signals();
+    // We need to do the above BEFORE we call InitModuleObjects
+    try
+    {
+        InitModuleObjects();
+    }
+    catch (IException *E)
+    {
+        EXCLOG(E);
+        E->Release();
+        return EXIT_FAILURE;
+    }
+    Owned<IProperties> globals = createProperties(true);
+    for (unsigned i=0; i<(unsigned)argc; i++)
+    {
+        if (stricmp(argv[i], "--help")==0 ||
+            stricmp(argv[i], "-h")==0)
+        {
+            topo_server_usage();
+            return EXIT_SUCCESS;
+        }
+        else if (streq(argv[i],"--daemon") || streq(argv[i],"-d")) {
+            if (daemon(1,0) || write_pidfile(argv[++i])) {
+                perror("Failed to daemonize");
+                return EXIT_FAILURE;
+            }
+        }
+        else
+            globals->loadProp(argv[i], true);  // We ignore unrecognized options for now
+    }
+    traceLevel = globals->getPropInt("--traceLevel", 1);
+    topoPort = globals->getPropInt("--port", TOPO_SERVER_PORT);
+    if (globals->getPropBool("--stdlog", traceLevel != 0))
+        queryStderrLogMsgHandler()->setMessageFields(MSGFIELD_time | MSGFIELD_milliTime | MSGFIELD_thread | MSGFIELD_prefix);
+    else
+        removeLog();
+    const char *logdir = globals->queryProp("--logdir");
```

Comment:
Would also be useful have options to configure the timeouts.
Should there also be a local .xml file to provides the defaults from the configuration?

Reply:
There's quite a bit to be done to productionalize the toposerver, I suspect. I'll add a subtask.

Reply:
https://track.hpccsystems.com/browse/HPCC-22492